File: game/scripts/src/modules/pathfinder/search/AStar.ts
import { SearchFunction, Pathfinder, PathNode } from '../types';
import { Heuristics } from '../core/Heuristics';

/**
 * A* pathfinding algorithm implementation
 * Combines the best of Dijkstra's algorithm and greedy best-first search
 * Based on zizouqi autochess project's A* implementation
 */
export class AStar {
    /**
     * A* search function
     * Based on zizouqi project's A* search logic
     */
    public static search: SearchFunction = (
        finder: Pathfinder,
        startNode: PathNode,
        endNode: PathNode,
        toClear: Set<PathNode>,
        tunnel: boolean = false
    ): PathNode | null => {
        const heuristic = finder.heuristic;
        
        // Initialize start node (based on zizouqi project's initialization)
        startNode.g = 0;
        startNode.h = heuristic(endNode.x - startNode.x, endNode.y - startNode.y);
        startNode.f = startNode.g + startNode.h;
        startNode.opened = true;
        toClear.add(startNode);

        // Use a simple array as open list for now (can be optimized with BinaryHeap)
        // Based on zizouqi project's open list management
        const openList: PathNode[] = [startNode];
        
        while (openList.length > 0) {
            // Find node with lowest F cost (based on zizouqi project's node selection)
            let currentIndex = 0;
            for (let i = 1; i < openList.length; i++) {
                if (openList[i].f < openList[currentIndex].f) {
                    currentIndex = i;
                }
            }
            
            const currentNode = openList.splice(currentIndex, 1)[0];
            currentNode.closed = true;
            
            // Check if we reached the goal (based on zizouqi project's goal check)
            if (currentNode === endNode) {
                return currentNode;
            }
            
            // Get neighbors (based on zizouqi project's neighbor expansion)
            const neighbors = finder.grid.getNeighbours(
                currentNode, 
                finder.walkable, 
                finder.allowDiagonal, 
                tunnel
            );
            
            for (const neighbor of neighbors) {
                if (neighbor.closed) {
                    continue;
                }
                
                toClear.add(neighbor);
                
                // Calculate tentative G cost (based on zizouqi project's cost calculation)
                const tentativeG = currentNode.g + this.calculateCost(currentNode, neighbor);
                
                if (!neighbor.opened || tentativeG < neighbor.g) {
                    // This path is better than any previous one
                    // Based on zizouqi project's path improvement logic
                    neighbor.parent = currentNode;
                    neighbor.g = tentativeG;
                    neighbor.h = heuristic(endNode.x - neighbor.x, endNode.y - neighbor.y);
                    neighbor.f = neighbor.g + neighbor.h;
                    
                    if (!neighbor.opened) {
                        openList.push(neighbor);
                        neighbor.opened = true;
                    }
                }
            }
        }
        
        // No path found
        return null;
    };

    /**
     * Calculate movement cost between two adjacent nodes
     * Based on zizouqi project's cost calculation system
     */
    private static calculateCost(from: PathNode, to: PathNode): number {
        const dx = Math.abs(to.x - from.x);
        const dy = Math.abs(to.y - from.y);
        
        if (dx === 0 || dy === 0) {
            // Orthogonal movement
            return 1;
        } else {
            // Diagonal movement
            return Math.SQRT2;
        }
    }
}

File: game/scripts/vscripts/modules/pathfinder/search/AStar.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local Set = ____lualib.Set
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["11"] = 9,["12"] = 9,["13"] = 9,["15"] = 9,["16"] = 93,["17"] = 94,["18"] = 95,["19"] = 97,["20"] = 99,["22"] = 102,["24"] = 93,["25"] = 14,["26"] = 19,["27"] = 19,["29"] = 21,["30"] = 24,["31"] = 25,["32"] = 26,["33"] = 27,["34"] = 28,["35"] = 32,["36"] = 34,["37"] = 36,["39"] = 37,["40"] = 37,["41"] = 38,["42"] = 39,["44"] = 37,["47"] = 43,["48"] = 44,["49"] = 47,["50"] = 48,["52"] = 52,["53"] = 59,["57"] = 60,["58"] = 61,["61"] = 64,["62"] = 67,["63"] = 69,["64"] = 72,["65"] = 73,["66"] = 74,["67"] = 75,["68"] = 77,["69"] = 78,["70"] = 79,["81"] = 86,["82"] = 14});
local ____exports = {}
--- A* pathfinding algorithm implementation
-- Combines the best of Dijkstra's algorithm and greedy best-first search
-- Based on zizouqi autochess project's A* implementation
____exports.AStar = __TS__Class()
local AStar = ____exports.AStar
AStar.name = "AStar"
function AStar.prototype.____constructor(self)
end
function AStar.calculateCost(self, from, to)
    local dx = math.abs(to.x - from.x)
    local dy = math.abs(to.y - from.y)
    if dx == 0 or dy == 0 then
        return 1
    else
        return 1.4142135623730951
    end
end
AStar.search = function(____, finder, startNode, endNode, toClear, tunnel)
    if tunnel == nil then
        tunnel = false
    end
    local heuristic = finder.heuristic
    startNode.g = 0
    startNode.h = heuristic(nil, endNode.x - startNode.x, endNode.y - startNode.y)
    startNode.f = startNode.g + startNode.h
    startNode.opened = true
    toClear:add(startNode)
    local openList = {startNode}
    while #openList > 0 do
        local currentIndex = 0
        do
            local i = 1
            while i < #openList do
                if openList[i + 1].f < openList[currentIndex + 1].f then
                    currentIndex = i
                end
                i = i + 1
            end
        end
        local currentNode = __TS__ArraySplice(openList, currentIndex, 1)[1]
        currentNode.closed = true
        if currentNode == endNode then
            return currentNode
        end
        local neighbors = finder.grid:getNeighbours(currentNode, finder.walkable, finder.allowDiagonal, tunnel)
        for ____, neighbor in ipairs(neighbors) do
            do
                local __continue12
                repeat
                    if neighbor.closed then
                        __continue12 = true
                        break
                    end
                    toClear:add(neighbor)
                    local tentativeG = currentNode.g + self:calculateCost(currentNode, neighbor)
                    if not neighbor.opened or tentativeG < neighbor.g then
                        neighbor.parent = currentNode
                        neighbor.g = tentativeG
                        neighbor.h = heuristic(nil, endNode.x - neighbor.x, endNode.y - neighbor.y)
                        neighbor.f = neighbor.g + neighbor.h
                        if not neighbor.opened then
                            openList[#openList + 1] = neighbor
                            neighbor.opened = true
                        end
                    end
                    __continue12 = true
                until true
                if not __continue12 then
                    break
                end
            end
        end
    end
    return nil
end
return ____exports
